junit-jupiter-api/src/main/java/org/junit/jupiter/api/io/TempDirFactory.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2015-2023 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * https://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junit.jupiter.api.io;
+
+import static org.apiguardian.api.API.Status.EXPERIMENTAL;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.attribute.FileAttribute;
+
+import org.apiguardian.api.API;
+import org.junit.platform.commons.util.Preconditions;
+import org.junit.platform.commons.util.ReflectionUtils;
+
+/**
+ * {@code TempDirFactory} defines the SPI for creating temporary directories
+ * programmatically.
+ *
+ * <p>A temporary directory factory is typically used to gain more control on
+ * the temporary directory creation, like defining the parent directory or even
+ * the file system that should be used.
+ *
+ * <p>Concrete implementations must have a <em>default constructor</em>.
+ *
+ * <p>A {@link TempDirFactory} can be configured <em>locally</em>
+ * for a test class field or method parameter via the {@link TempDir @TempDir}
+ * annotation.
+ *
+ * @since 5.9
+ * @see TempDir @TempDir
+ */
+@FunctionalInterface
+@API(status = EXPERIMENTAL, since = "5.9")
+public interface TempDirFactory {
+
+	/**
+	 * Create a new temporary directory, using the given prefix to generate its name.
+	 * Depending on the implementation, the resulting {@code Path} may or may not be
+	 * associated with the default {@code FileSystem}.
+	 *
+	 * @param prefix the prefix string that can be used in generating the directory's name; never {@code null} or blank
+	 * @return the path to the newly created directory that did not exist before this method was invoked; never {@code null}
+	 * @throws Exception in case of failures
+	 */
+	Path createTempDirectory(String prefix) throws Exception;
```

Comment:
could we have something such the current test class name in order to create directory depending on the name.
Use case is to debug content of created temp directories and/or files in case of test failures.

Reply:
@scordio do you need a PR for this enhancement

Reply:
Hi @olamy, I'm slowly catching up with the open points I have in this PR.

I'm ok with your proposal. I would also wait for feedback from the team before going in this direction as it requires a bit of refactoring to get the context as a method parameter. Once we get it, I'm happy to handle those changes.

Reply:
Triggered by https://github.com/junit-team/junit5/pull/2958#issuecomment-145341381, I replaced the `prefix` parameter of `createTempDirectory` with the `ExtensionContext`.